src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+                    // Already in the visible list
+                    return;
+                }
+            }
+
+            _visibleFrames = _visibleFrames.Add(new FrameListener(this, frame));
+
+            this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        private void RemoveFrame(FrameListener frame)
+        {
+            _visibleFrames = _visibleFrames.Remove(frame);
+
+            this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+        public int OnSelectionChanged(IVsHierarchy pHierOld, [ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSITEMID")]uint itemidOld, IVsMultiItemSelect pMISOld, ISelectionContainer pSCOld, IVsHierarchy pHierNew, [ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSITEMID")]uint itemidNew, IVsMultiItemSelect pMISNew, ISelectionContainer pSCNew)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public int OnElementValueChanged([ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSSELELEMID")]uint elementid, object varValueOld, object varValueNew)
+        {
+            AssertIsForeground();
+
+            if (elementid == (uint)VSConstants.VSSELELEMID.SEID_DocumentFrame)
+            {
+                // Remember the newly activated frame so it can be read from another thread.
+                _activeFrame = varValueNew as IVsWindowFrame;
+
+                if (_activeFrame != null)
+                {
+                    // This might be the first time we've ever seen this IVsWindowFrame, so add it to our tracking too
+                    TrackWindowFrame(_activeFrame);
+                }
+
+                this.DocumentsChanged?.Invoke(this, EventArgs.Empty);
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnCmdUIContextChanged([ComAliasName("Microsoft.VisualStudio.Shell.Interop.VSCOOKIE")]uint dwCmdUICookie, [ComAliasName("Microsoft.VisualStudio.OLE.Interop.BOOL")]int fActive)
+        {
+            return VSConstants.E_NOTIMPL;
+        }
+
+        public event EventHandler<EventArgs> NonRoslynBufferTextChanged;
+
+        /// <summary>
+        /// Listens to frame notifications for a visible frame. When the frame becomes invisible or closes,
+        /// then it automatically disconnects.
+        /// </summary>
+        private class FrameListener : IVsWindowFrameNotify, IVsWindowFrameNotify2
+        {
+            public readonly IVsWindowFrame Frame;
+
+            private readonly VisualStudioActiveDocumentTracker _documentTracker;
+            private readonly uint _frameEventsCookie;
+
+            private readonly ITextBuffer _textBuffer;
+
+            public FrameListener(VisualStudioActiveDocumentTracker service, IVsWindowFrame frame)
+            {
+                _documentTracker = service;
+                _documentTracker.AssertIsForeground();
+
+                this.Frame = frame;
+
+                ((IVsWindowFrame2)frame).Advise(this, out _frameEventsCookie);
```

Comment:
can this fail in any way?   c

Reply:
Yes, it looks like the old code let that happen, and then just checked if the cookie was null before unsubscribing. Restored the check (since you called that out elsewhere too)